aws/provider_test.go

```diff
@@ -1190,6 +1183,44 @@ func testAccPreCheckSkipError(err error) bool {
 	return false
 }
 
+type testSweepResource struct {
+	d        *schema.ResourceData
+	meta     interface{}
+	resource *schema.Resource
+}
+
+func NewTestSweepResource(resource *schema.Resource, d *schema.ResourceData, meta interface{}) *testSweepResource {
+	return &testSweepResource{
+		d:        d,
+		meta:     meta,
+		resource: resource,
+	}
+}
+
+func testSweepResourceOrchestrator(sweepResources []*testSweepResource) error {
+	var wg sync.WaitGroup
+	var errors *multierror.Error
+	var mutex = &sync.Mutex{}
+
+	for _, sweepResource := range sweepResources {
+		wg.Add(1)
+
+		go func(sweepResource *testSweepResource) {
+			defer wg.Done()
+
+			if err := testAccDeleteResource(sweepResource.resource, sweepResource.d, sweepResource.meta); err != nil {
+				mutex.Lock()
+				errors = multierror.Append(errors, err)
+				mutex.Unlock()
+			}
+		}(sweepResource)
+	}
+
+	wg.Wait()
+
+	return errors.ErrorOrNil()
+}
```

Comment:
Using [`multierror.Group`](https://pkg.go.dev/github.com/hashicorp/go-multierror#Group), we can move all the concurrency handling upstream via:

```suggestion
func testSweepResourceOrchestrator(sweepResources []*testSweepResource) error {
	var g multierror.Group

	for _, sweepResource := range sweepResources {
		sweepResource := sweepResource // https://golang.org/doc/faq#closures_and_goroutines
		g.Go(func() error {
			return testAccDeleteResource(sweepResource.resource, sweepResource.d, sweepResource.meta)
		})
	}

	return g.Wait().ErrorOrNil()
}
```

Verified by creating two database instances via:

```terraform
terraform {
  required_version = "0.14.9"
  required_providers {
    aws = {
      source  = "hashicorp/aws"
      version = "3.35.0"
    }
  }
}

provider "aws" {
  region = "us-east-2"
}

data "aws_rds_orderable_db_instance" "test" {
  engine         = "mysql"
  engine_version = "5.6.35"
  license_model  = "general-public-license"
  storage_type   = "standard"

  preferred_instance_classes = ["db.t3.micro", "db.t2.micro", "db.t2.medium"]
}

resource "aws_db_instance" "example" {
  for_each = toset(["one", "two"])

  allocated_storage    = 10
  engine               = data.aws_rds_orderable_db_instance.test.engine
  engine_version       = data.aws_rds_orderable_db_instance.test.engine_version
  instance_class       = data.aws_rds_orderable_db_instance.test.instance_class
  name                 = "bfladtesting${each.value}"
  parameter_group_name = "default.mysql5.6"
  password             = "barbarbarbar"
  skip_final_snapshot  = true
  username             = "foo"
}
```

```console
$ terraform apply
...
Apply complete! Resources: 2 added, 0 changed, 0 destroyed.

$ go test ./aws -v -timeout=10h -sweep-allow-failures -sweep=us-east-2 -sweep-run=aws_db_instance
2021/04/01 20:32:13 [DEBUG] Running Sweepers for region (us-east-2):
2021/04/01 20:32:13 [DEBUG] Running Sweeper (aws_db_instance) in region (us-east-2)
2021/04/01 20:32:14 [DEBUG] DB Instance destroy: terraform-20210402002337556200000002
2021/04/01 20:32:14 [DEBUG] DB Instance destroy: terraform-20210402002337556200000001
2021/04/01 20:32:14 [DEBUG] DB Instance destroy configuration: {
  DBInstanceIdentifier: "terraform-20210402002337556200000002",
  DeleteAutomatedBackups: false,
  SkipFinalSnapshot: true
}
2021/04/01 20:32:14 [DEBUG] DB Instance destroy configuration: {
  DBInstanceIdentifier: "terraform-20210402002337556200000001",
  DeleteAutomatedBackups: false,
  SkipFinalSnapshot: true
}
2021/04/01 20:32:14 [INFO] Waiting for DB Instance to be destroyed
2021/04/01 20:32:14 [INFO] Waiting for DB Instance to be destroyed
2021/04/01 20:32:44 [DEBUG] DB Instance describe configuration: {
  DBInstanceIdentifier: "terraform-20210402002337556200000001"
}
2021/04/01 20:32:44 [DEBUG] DB Instance describe configuration: {
  DBInstanceIdentifier: "terraform-20210402002337556200000002"
}
2021/04/01 20:32:44 [DEBUG] DB Instance status for instance terraform-20210402002337556200000001: deleting
2021/04/01 20:32:44 [DEBUG] DB Instance status for instance terraform-20210402002337556200000002: deleting
2021/04/01 20:32:54 [DEBUG] DB Instance describe configuration: {
  DBInstanceIdentifier: "terraform-20210402002337556200000001"
}
2021/04/01 20:32:54 [DEBUG] DB Instance describe configuration: {
  DBInstanceIdentifier: "terraform-20210402002337556200000002"
}
2021/04/01 20:32:54 [DEBUG] DB Instance status for instance terraform-20210402002337556200000001: deleting
2021/04/01 20:32:54 [DEBUG] DB Instance status for instance terraform-20210402002337556200000002: deleting
...
2021/04/01 20:35:48 [DEBUG] DB Instance describe configuration: {
  DBInstanceIdentifier: "terraform-20210402002337556200000001"
}
2021/04/01 20:35:48 [DEBUG] DB Instance status for instance terraform-20210402002337556200000001: deleting
2021/04/01 20:35:58 [DEBUG] DB Instance describe configuration: {
  DBInstanceIdentifier: "terraform-20210402002337556200000001"
}
2021/04/01 20:35:58 [DEBUG] DB Instance status for instance terraform-20210402002337556200000001: deleting
2021/04/01 20:36:08 [DEBUG] DB Instance describe configuration: {
  DBInstanceIdentifier: "terraform-20210402002337556200000001"
}
2021/04/01 20:36:08 Sweeper Tests ran successfully:
	- aws_db_instance
ok  	github.com/terraform-providers/terraform-provider-aws/aws	237.624s
```

Reply:
It's a strange combination of error collecting and concurrency. It violates the single responsibility principle and might be confusing to future travelers that the concurrency comes from `multierror`. But, Nomad copied it from `golang.org/x` and it's good enough for Mitchell, so my concerns are of no concern.